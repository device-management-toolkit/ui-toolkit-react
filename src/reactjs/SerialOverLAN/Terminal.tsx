/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/

import React from 'react'
import Style from 'styled-components'
import { isFalsy } from '../shared/Utilities'

const TerminalContainer = Style.div`
   display:block;
   text-align:center;
`
const XTerm = Style.div`
   display:inline-block;
`

export interface IPropTerminal {
  handleKeyPress: any
  xterm: any
  handleKeyDownPress: any
}

class Term extends React.Component<IPropTerminal> {
  componentDidMount (): void {
    const element = document.getElementById('xterm') ?? ''
    const { xterm, handleKeyPress } = this.props
    if (isFalsy(element)) {
      xterm.open(element)
      xterm.onData(data => handleKeyPress(data))
      xterm.attachCustomKeyEventHandler(e => {
        e.stopPropagation()
        e.preventDefault()
        if (isFalsy(e.ctrlKey) && isFalsy(e.shiftKey) && (e.keyCode === 67)) {
          return navigator.clipboard.writeText(xterm.getSelection())
        } else if (isFalsy(e.ctrlKey) && isFalsy(e.shiftKey) && (e.keyCode === 86)) {
          return navigator.clipboard.readText()
            .then(text => handleKeyPress(text)
            )
        } else if (e.code === 'Space') {
          return handleKeyPress(e.key)
        }
      })
    }
  }

  render (): React.ReactNode {
    return (
      <TerminalContainer>
        <XTerm id="xterm" />
      </TerminalContainer>
    )
  }
}

export default Term
