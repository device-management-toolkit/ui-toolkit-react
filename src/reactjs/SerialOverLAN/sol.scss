/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/
.sol-header {
    display: flex;
}
.power-options {
	display: block;
	position: absolute;
	right: 1%;
}
