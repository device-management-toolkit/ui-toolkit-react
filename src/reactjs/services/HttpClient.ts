/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/

import { prepareHeaders } from '../shared/Utilities'

export const HttpClient = {
  async get (url, options = {}): Promise<any> {
    const jsonHeaders = prepareHeaders()
    return await HttpClient.fetch(url, {
      ...options,
      method: 'GET',
      headers: jsonHeaders
    })
  },

  async post (url, body = {}, options = {}): Promise<any> {
    const jsonHeaders = prepareHeaders()
    return await HttpClient.fetch(url, {
      ...options,
      method: 'POST',
      body: body,
      headers: jsonHeaders
    })
  },

  async delete (url, options = {}): Promise<any> {
    const jsonHeaders = prepareHeaders()
    return await HttpClient.fetch(url, {
      ...options,
      method: 'DELETE',
      headers: jsonHeaders
    })
  },

  async fetch (url, options = {}): Promise<any> {
    (options as any).mode = 'cors';
    (options as any).credentials = 'include';
    (options as any).withCredentials = true
    return await fetch(url, options).then(async result => await result.json())
  }
}
