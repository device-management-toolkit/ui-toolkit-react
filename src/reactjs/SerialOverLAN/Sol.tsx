/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/

import React from 'react'
import { AmtTerminal, AMTRedirector, Protocol, ConsoleLogger, LogLevel, TerminalDataProcessor } from '@open-amt-cloud-toolkit/ui-toolkit/core'
import Style from 'styled-components'

import { Terminal } from 'xterm'
import Term from './Terminal'
import 'xterm/css/xterm.css'
import './sol.scss'

const StyledDiv = Style.div`
display : inline-block;
padding : 0px 5px;
`

const HeaderStrip = Style.div`
background-color: darkgray;
padding: 5px;
font-size: 13px;
text-align: center;
`

export interface SOLProps {
  deviceId: string | null
  mpsServer: string | null
  autoConnect?: boolean
  authToken: string
}

export interface SOLStates {
  isConnected: boolean
  SOLstate: number
  powerState: number
  showSuccess: boolean
  message: string
  isSelected: boolean
  type: string
  solNotEnabled: string
  deviceOnSleep: string
  isPowerStateLoaded: boolean
}

/** container class for SOL */
export class Sol extends React.Component<SOLProps, SOLStates> {
  redirector: any
  terminal: any
  logger: any
  dataProcessor: any
  callback: any
  term: any
  fr: FileReader
  constructor (props: SOLProps) {
    super(props)
    this.logger = new ConsoleLogger(LogLevel.ERROR)
    this.state = {
      isConnected: false,
      SOLstate: 0,
      powerState: 0,
      showSuccess: false,
      message: '',
      isSelected: true,
      type: '',
      solNotEnabled: '',
      deviceOnSleep: '',
      isPowerStateLoaded: false
    }
  }

  init = (): void => {
    const server: string = this.props.mpsServer != null ? this.props.mpsServer.replace('http', 'ws') : ''
    const deviceUuid: string = this.props.deviceId != null ? this.props.deviceId : ''
    this.terminal = new AmtTerminal()
    this.redirector = new AMTRedirector(
      this.logger,
      Protocol.SOL,
      new FileReader(),
      deviceUuid,
      16994,
      '',
      '',
      0,
      0,
      this.props.authToken,
      `${server}/relay`
    )
    this.dataProcessor = new TerminalDataProcessor(this.terminal)
    this.terminal.onSend = this.redirector.send.bind(this.redirector)
    this.redirector.onNewState = this.terminal.StateChange.bind(this.terminal)
    this.redirector.onStateChanged = this.onTerminalStateChange.bind(this)
    this.redirector.onProcessData = this.dataProcessor.processData.bind(this.dataProcessor)
    this.dataProcessor.processDataToXterm = this.handleWriteToXterm.bind(this)
    this.dataProcessor.clearTerminal = this.handleClearTerminal.bind(this)
    this.term = new Terminal({
      cursorStyle: 'block',
      fontWeight: 'bold',
      rows: 30,
      cols: 100
    })
  }

  cleanUp = (): void => {
    this.terminal = null
    this.redirector = null
    this.dataProcessor = null
    this.term = null
  }

  componentDidMount (): void {
    this.init()
  }

  /** write the processed data from webscoket in to xterm */

  handleWriteToXterm = (str): any => this.term.write(str)

  handleClearTerminal = (): any => this.term.reset()

  /** capture the data on xterm key press */
  handleKeyPress = (domEvent): any => this.terminal.TermSendKeys(domEvent)

  handleKeyDownPress = (domEvent): any => this.terminal.handleKeyDownEvents(domEvent)

  startSOL = (): void => {
    if (typeof this.redirector !== 'undefined') {
      this.redirector.start(WebSocket)
    }
  }

  stopSOL = (): void => {
    if (typeof this.redirector !== 'undefined') {
      this.redirector.stop()
    }
    this.handleClearTerminal()
    this.cleanUp()
    this.init()
  }

  handleSOLConnect = (e): void => {
    e.persist()
    if (this.state.SOLstate === 0) {
      this.startSOL()
    } else {
      this.stopSOL()
    }
  }

  onTerminalStateChange = (redirector, state: number): void => this.setState({ SOLstate: state })

  /** callback functions from child components to update the state values */
  handleFeatureStatus = (value): void => {
    this.setState({
      solNotEnabled: value
    })
  }

  getSOLState = (): any => this.state.SOLstate === 3 ? 2 : 0

  render (): React.ReactNode {
    const { SOLstate } = this.state
    return (
      <React.Fragment>
        <HeaderStrip>
          <StyledDiv>
            <button onClick={this.handleSOLConnect}>{SOLstate === 3 ? 'Disconnect' : 'Connect'}</button>
          </StyledDiv>
        </HeaderStrip>
        {SOLstate === 3 && this.term && <Term handleKeyPress={this.handleKeyPress} handleKeyDownPress={this.handleKeyDownPress} xterm={this.term} />}
      </React.Fragment>
    )
  }
}
