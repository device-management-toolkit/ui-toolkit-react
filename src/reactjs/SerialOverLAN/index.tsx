/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/

import React from 'react'
import { render } from 'react-dom'
import { Sol } from './Sol'
import i18n from '../../i18n'
// Get browser language
i18n.changeLanguage(navigator.language).catch(() => console.info('error occured'))

const url = new URL(window.location.href)
const params = new URLSearchParams(url.search)

render(<Sol deviceId={params.get('deviceId')} authToken="authToken" mpsServer={params.get('mpsServer')} />, document.getElementById('sol'))
