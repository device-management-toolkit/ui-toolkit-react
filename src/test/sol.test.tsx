/*********************************************************************
* Copyright (c) Intel Corporation 2019
* SPDX-License-Identifier: Apache-2.0
**********************************************************************/
import { Sol, SOLProps } from '../reactjs/SerialOverLAN/Sol'
import { HttpClient } from '../reactjs/services/HttpClient'
import { AMTRedirector1, Protocol } from './helper/amtredirector1'
import { Mockeventpersist } from './helper/mockeventpersist'

import React from 'react'
import { shallow } from 'enzyme'
import { Terminal } from './helper/terminal'
import { AmtTerminal } from '@open-amt-cloud-toolkit/ui-toolkit/core'
// import { FileReader } from '../core/FileReader'

const solprop: SOLProps = {
  deviceId: 'acfae359-be7b-4861-8e0c-54b20389bb68',
  mpsServer: 'https://localhost:9300',
  authToken: 'authToken'
}

describe('Test Sol class', () => {
  // const fr: FileReader = {
  //   onload: jest.fn(),
  //   onloadend: jest.fn(),
  //   readAsArrayBuffer: jest.fn(),
  //   readAsBinaryString: jest.fn()
  // }
  it('Test Sol render', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    // shallow
    const sol = shallow(<Sol {...solprop} />)

    // Output
    expect(sol).toMatchSnapshot()
  })

  it('Test handlePowerOptions with failure', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }
    HttpClient.post = jest.fn(async function (arg1, arg2, arg3) {
      return await new Promise((resolve, reject) => {
        resolve({
          Body: {
            ReturnValueStr: 'FAIL'
          }
        })
      })
    })

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol
    console.info('myinstance', myInstance)

    // call the function and check the expected output
    // return myInstance.handlePowerOptions(ev).then(data => expect(myInstance.state.message).toEqual("Sorry! there was some technical difficulties"));
  })

  it('Test onTerminalStateChange', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    // call the function and check the expected output
    myInstance.onTerminalStateChange(null, 2)
    expect(myInstance.state.SOLstate).toBe(2)
  })

  it('Test startSOL', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol
    myInstance.redirector = new AMTRedirector1(myInstance.logger,
      Protocol.SOL,
      new FileReader(),
      'abcd-efgh-ijkl-mnop',
      16994,
      '',
      '',
      0,
      0,
      'authToken',
      '1.2.3.4:9876' + '/relay')

    // call the function and check the expected output
    myInstance.startSOL()
    expect(myInstance.redirector.startvariable).toBe(5)
  })

  it('Test stopSOL', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol
    myInstance.redirector = new AMTRedirector1(myInstance.logger,
      Protocol.SOL,
      new FileReader(),
      'abcd-efgh-ijkl-mnop',
      16994,
      '',
      '',
      0,
      0,
      'authToken',
      '1.2.3.4:9876' + '/relay')

    // call the function and check the expected output
    myInstance.startSOL()
    myInstance.stopSOL()
    expect(myInstance.state.SOLstate).toBe(0)
  })

  it('Test onTerminalStateChange', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    // call the function and check the expected output
    myInstance.onTerminalStateChange(null, 4)
    expect(myInstance.state.SOLstate).toBe(4)
  })

  it('Test handleSOLConnect to start', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }
    const e = new Mockeventpersist()

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    myInstance.redirector = new AMTRedirector1(myInstance.logger,
      Protocol.SOL,
      new FileReader(),
      'abcd-efgh-ijkl-mnop',
      16994,
      '',
      '',
      0,
      0,
      'authToken',
      '1.2.3.4:9876' + '/relay')

    // call the function and check the expected output
    myInstance.onTerminalStateChange(null, 0)
    myInstance.handleSOLConnect(e)
    expect(myInstance.redirector.startvariable).toBe(5)
  })

  it('Test handleSOLConnect to stop', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }
    const e = new Mockeventpersist()

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    myInstance.redirector = new AMTRedirector1(myInstance.logger,
      Protocol.SOL,
      new FileReader(),
      'abcd-efgh-ijkl-mnop',
      16994,
      '',
      '',
      0,
      0,
      'authToken',
      '1.2.3.4:9876' + '/relay')

    // call the function and check the expected output
    myInstance.onTerminalStateChange(null, 1)
    myInstance.handleSOLConnect(e)
    expect(myInstance.state.isConnected).toBe(false)
    expect(myInstance.state.SOLstate).toBe(1)
  })

  it('Test handleWriteToXterm', () => {
    // Initialization of ConnectProps
    const solprop: SOLProps = {
      deviceId: 'abcd-efh-ijkl-mnop',
      mpsServer: '1.2.3.4:1234',
      authToken: 'authToken'
    }

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol
    myInstance.term = new Terminal()

    // call the function and check the expected output
    myInstance.handleWriteToXterm('abcdef')
    expect(myInstance.term.writestring).toBe('abcdef')
  })

  it('Test handleClearTerminal', () => {
    // Initialization of ConnectProps

    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol
    myInstance.term = new Terminal()

    // call the function and check the expected output
    myInstance.handleClearTerminal()
    expect(myInstance.term.resetvalue).toBe(1)
  })

  it('should call the handle key events on trigger ', () => {
    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    myInstance.terminal = new AmtTerminal()
    myInstance.terminal.TermSendKeys = jest.fn((domEvent) => { })
    myInstance.terminal.handleKeyDownEvents = jest.fn((domEvent) => { })

    const domEvent = {
      target: {
        value: ''
      }
    }
    myInstance.handleKeyPress(domEvent)
    myInstance.handleKeyDownPress(domEvent)
    expect(myInstance.terminal.TermSendKeys).toHaveBeenCalled()
  })

  it('should update the component from the child component on the feature and power status', () => {
    const sol = shallow(<Sol {...solprop} />)
    const myInstance = sol.instance() as Sol

    myInstance.handleFeatureStatus('enabled')
    expect(sol.state('solNotEnabled')).toEqual('enabled')

    myInstance.handleFeatureStatus('notEnabled')
    expect(sol.state('solNotEnabled')).toEqual('notEnabled')

    myInstance.handleFeatureStatus('failed')
    expect(sol.state('solNotEnabled')).toEqual('failed')
  })
})
