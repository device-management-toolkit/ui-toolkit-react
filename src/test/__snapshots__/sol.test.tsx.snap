// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Sol class Test Sol render 1`] = `ShallowWrapper {}`;
